docs/madr/decisions/057-rework-mesh-routes-matching.md

```diff
@@ -0,0 +1,895 @@
+      http:
+        requestTimeout: 5s
+  - targetRef:
+      kind: MeshHTTPRoute
+      name: slow-backend-route
+      namespace: backend-ns
+    default:
+      http:
+        requestTimeout: 15s
+```
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTimeout
+metadata:
+  name: timeout-on-backend
+  namespace: frontend-ns
+  labels:
+    kuma.io/mesh: default
+spec:
+  targetRef:
+    kind: MeshService
+    name: backend
+    namespace: backend-ns
+  default:
+    http:
+      requestTimeout: 5s
+---
+apiVersion: kuma.io/v1alpha1
+kind: MeshTimeout
+metadata:
+  name: timeout-on-backend
+  namespace: frontend-ns
+  labels:
+    kuma.io/mesh: default
+spec:
+  targetRef:
+    kind: MeshHTTPRoute
+    name: slow-backend-route
+    namespace: backend-ns
+  default:
+    http:
+      requestTimeout: 15s
+```
+
+## Mesh*Route in spec.to[].targetRef design
+
+The design is grounded in multiple recent MADRs around improvements to policies: 
+
+- [MeshService policy matching](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/046-meshservice-policy-matching.md)
+- [Policy on Namespace](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/052-policy-on-namespace.md) - which introduced consumer/producer policies
+- [Policy matching with real resources](https://github.com/kumahq/kuma/pull/10726/files)
+
+### API
+
+New API for policy that selects `Mesh*Route`
+
+```yaml
+type: MeshTimeout
+name: timeouts-on-route
+mesh: default
+spec:
+  targetRef:
+    kind: Mesh | MeshSubset # either Mesh, MeshSubset or MeshGateway
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute | MeshTCPRoute # either MeshHTTPRoute or MeshTCPRoute
+        name: route-to-backend
+      default:
+        http:
+          requestTimeout: 5s
+          streamIdleTimeout: 1h
+```
+
+specifying `Mesh` in `spec.targetRef` will be optional (this probably will be most used case). Minimal policy, targeting
+`MeshHTTPRoute` will look like this:
+
+```yaml
+type: MeshTimeout
+name: timeouts-on-route
+mesh: default
+spec:
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute
+        name: route-to-backend
+      default:
+        http:
+          requestTimeout: 5s
+          streamIdleTimeout: 1h
+```
+
+To introduce this change we need to deprecate specifying `Mesh*Route` in `spec.targetRef`.
+
+#### Additional validation
+
+When referencing Mesh*Route we should validate policy configuration to make sure user is applying only configuration 
+that can be applied on route.
+
+### Affected policies
+
+At the moment we can only target `Mesh*Routes` in MeshTimeout policy. This MADR is based on MeshTimeout policy.
+Looking at what can be configured at Envoy route we can also apply this to our MeshRetry policy. Rate limit is also configurable on route,
+but we are configuring rate limit on inbound traffic, which does not apply to this MADR. `MeshLoadBalancingStrategy` also
+configures hashing per route and we should probably add support for routes in the future
```

Comment:
This paragraph is hard to follow. I'd go with something more like a list.

Here are the policies that may be configurable on routes:

- `MeshTimeout`: can be configured per route. However, not all parameters make sense (.e.g: tcp timeouts). Validation will be added to reject policies with `spec.to[].targetRef.kind: MeshHTTPRoute` which configs that can't be configured. What about inspect policies and merging? Will we see TCP timeouts?
- `MeshLoadBalancingStrategy`: can be configured per route. Like `MeshTimeout` all configurations do not make sense and will need to be validated.
- `MeshRetry` : can be configured per route. Like `MeshTimeout` all configurations do not make sense and will need to be validated.
- `MeshRateLimit`: while it's possible to set rate limit on a route. We currently only set rate limits on the inbound so supporting `spec.to[].target.kind: MeshHTTPRoute` doesn't make sense.

What about:
- `MeshAccessLog`: Seems like we should be able to support it just for a route no?

Reply:
rewritten this section as suggested. As for:
> What about inspect policies and merging? Will we see TCP timeouts?

@lobkovilya correctly me if I am wrong but with current merging we will get all the information not only route specific, right?

Reply:
Yes, with the current implementation of ResourceRules we consider MeshHTTPRoutes included in MeshServices, so the final configuration for route is a product of merging:
```
meshservice:mesh/mesh-1:name/backend:
  conf: $ms_specific_conf
meshhttproute:mesh/mesh-1:name/route-to-backend:
  conf: merge($ms_specific_conf, $route_specific_conf)
```

Reply:
After discussion with @lobkovilya we've decided that we will change merging mechanism to not merge confs for different resources types, as it was now. So we won't include Mesh or MeshService conf in Mesh*Route conf. So in case of inspecting policies we will see only config for a route without defaults from Mesh or MeshService, as they could not be applicable on route itself